src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7363,9 +7363,19 @@ private BoundExpression ConvertToArrayIndex(BoundExpression index, SyntaxNode no
 
             var result =
                 TryImplicitConversionToArrayIndex(index, SpecialType.System_Int32, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt32, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_Int64, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt64, node, diagnostics);
+                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt32, node, diagnostics);
+
+            if (result is null && allowNIntAndNUInt)
```

Comment:
This is done as `nint`/`nuint` are only valid for single-dimensional arrays and pointers.

This PR only handles it for single-dimensional arrays as handling for pointers requires additional logic adding support for `PointerAndNInt` type operators which was going to be more involved.

Reply:
Also, notably, it could become valid for multidimensional arrays but would require new APIs in S.P.Corelib exposed on System.Array